### DISCUSSION TITLE
2D time-dependent heat equation with intermittent initial conditions

### DISCUSSION THREAD

**Discussion by MINE0126** [lululxvi/deepxde#874] [Q&A] [answered]
Dear Prof.Lu,
I'm trying to solve a 2D time-dependent heat equation with intermittent initial conditions using DEEPXDE. 
### the equation and boundary conditions:
The k is simplified and assumed coefficient
![2-D](https://user-images.githubusercontent.com/107633173/187392614-e1c56dfb-cb44-415f-8fc7-da17076b3f8c.png)
### my code:
```
import deepxde as dde
import matplotlib.pyplot as plt
import numpy as np
from deepxde.backend import tf

end_time = 0.2
def pde(x,y):
    dy_xx = dde.grad.hessian(y, x, i=0,j=0)
    dy_yy = dde.grad.hessian(y, x, i=1,j=1)
    dy_t = dde.grad.jacobian(y, x, j=2)
    k=0.5
    return ((dy_t) - (k * (dy_xx + dy_yy)))


def l_boundary(x,on_boundary):
    return on_boundary and np.isclose(x[1],0.8) and np.all([x[0]>=0,x[0]<=0.8])
def r_boundary(x,on_boundary):
    return on_boundary and np.isclose(x[1],0.8) and np.all([x[0]>=1.6,x[0]<=2.4])
def boundary_1(x,on_boundary):
    return on_boundary and np.isclose(x[0],0)
def boundary_2(x,on_boundary):
    return on_boundary and np.isclose(x[1],0)
def boundary_3(x,on_boundary):
    return on_boundary and np.isclose(x[0],2.4)
def boundary_4(x,on_boundary):
    return on_boundary and np.isclose(x[1],0.8) and np.all([x[0]>=0.8,x[0]<=1.6])

def boundary_t_init(x, on_initial):
    return np.isclose(x[2],0)
#def boundary_initial(X, on_initial):
#    x,y,t = X
#    return on_initial and np.isclose(t, 0)

def func_left(x):
    return 0.8

def func_right(x):
    return 0.2
def func_zeros(x):
    return np.zeros((len(x),1))

geom = dde.geometry.Rectangle(xmin=[0, 0], xmax=[2.4, 0.8])
timedomain = dde.geometry.TimeDomain(0, end_time)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)

bc_l = dde.DirichletBC(geomtime, func_left, l_boundary)
bc_r = dde.DirichletBC(geomtime, func_right, r_boundary)
bc_1 = dde.NeumannBC(geomtime, func_zeros, boundary_1)
bc_2 = dde.NeumannBC(geomtime, func_zeros, boundary_2)
bc_3 = dde.NeumannBC(geomtime, func_zeros, boundary_3)
bc_4 = dde.NeumannBC(geomtime, func_zeros, boundary_4)
ic = dde.IC(geomtime, func_zeros, boundary_initial)

data = dde.data.TimePDE(
    geomtime, pde, [bc_l, bc_r,bc_1,bc_2,bc_3,bc_4,ic], num_domain=500, num_boundary=100,  num_initial=600)

net = dde.maps.FNN([3] + [60] * 5 + [1], "tanh", "Glorot uniform")
net.apply_output_transform(lambda x, y: abs(y))
model = dde.Model(data, net)

model.compile("adam", lr=0.001)
losshistory, trainstate = model.train(epochs=20000)
model.compile("L-BFGS-B")
losshistory, train_state = model.train(epochs = 20000)
dde.saveplot(losshistory, trainstate, issave=True, isplot=True)
```

### The results of DEEPXDE
https://drive.google.com/file/d/1PR3oIcvEhep5VadB6nhvtZvPnzHtf3lZ/view?usp=sharing
when t=0.1, the results of pinn is 
![ex02-pinn(t=0 1)](https://user-images.githubusercontent.com/107633173/187410804-dd9a61a2-d144-4ccc-b7fb-8583fef455e4.png)

### The results of FEM
https://drive.google.com/file/d/17cr1d9GednveazPZwZWYXj4_kQzkp3pf/view?usp=sharing
when t=0.1, the results of fem is
![ex02-FEM(t=0 1)](https://user-images.githubusercontent.com/107633173/187411156-4e8882af-2810-4c4d-a5d7-0235abd0bbbe.png)

### My quetion
When t =0, T=0 (the initial conditions). Whether the initial conditions are correctly in my code? Maybe I have other mistakes, welcome to correct.
Thank you very much!

**Comment by lululxvi** [comment, marked as answer]
IC looks OK. Or you can plot the PINN prediction at t=0 to verify.

**Reply by MINE0126** [reply]
Hi, @lululxvi 
Thank you for your reply. 
https://github.com/lululxvi/deepxde/issues/462#issuecomment-1002830981
The specific operation of scaling the network is: Encrypt the number of training points within the domain and on the boundaries?
Looking forward to your reply! Thanks again.

**Reply by lululxvi** [reply]
What do you mean by "Encrypt"?

**Reply by MINE0126** [reply]
So embarrassing!
In https://github.com/lululxvi/deepxde/issues/462#issuecomment-1002830981, the specific operation of scaling the network is: increase the number of training points within the domain and on the boundaries, right?
I am really sorry! Looking forward to your reply!

**Reply by MINE0126** [reply]
I set the hard constraint as follows
```
def transform(x,y):
	if x[1]==0.8 and 0.0<=x[0]<=0.8:
		y==0.8
	elif x[1]==0.8 and 1.6<=x[0]<=2.4:
		y==0.2
	else:
		y==y
	return y
net.apply_output_transform(transform)
```
Is that right? Thanks.

**Reply by lululxvi** [reply]
You should use tensorflow/pytorch functions, not `if`, `else`
